### PR TITLE
RollingStock: finish investment phase; add ipo phase

### PR DESCRIPTION
**Pull request by roseundy**
Add buy/sell stocks to the Investment phase/round (along with price changes)
Implement the IPO phase/round

Common file edits:
- UI::EntityList - use company short names instead of long names
- UI::Par - don't call entity.cash unless needed
- UI::GamePage - use the StockRound UI if @round.stock? is true
- Engine::Company - allow companies to pass! and unpass!
- Engine::SharePool - allow silencing of log messages on buy_shares

### CODE REVIEW

File: lib/engine/game/g_rolling_stock/game.rb [hunk 1]
```diff
@@ -188,6 +247,57 @@ def update_offering(company)
...
+
+        def ipo_companies
+          @companies.select { |c| c.owner&.player? && c.owner != @foreign_investor }.sort_by(&:value)
+        end
+
+        def available_par_prices(company)
+          PAR_PRICES[@company_stars[company.sym]].map { |p| lookup_price(p) }.select { |p| p.corporations.empty? }
+        end
+
+        def lookup_price(price)
```
Comment (tobymao) [inline lib/engine/game/g_rolling_stock/game.rb:263]: this looks really inefficient, you're doing n^2 operations every time you  do lookup_price?

File: lib/engine/game/g_rolling_stock/game.rb [hunk 2]
```diff
@@ -148,33 +180,60 @@ def reorder_by_cash
...
 
         # income
-        def phase5; end
+        def phase5
+          @log << "-- Turn #{@turn}, Phase 5 - Income --"
+          (@players + [@foreign_investor] + @corporations).each do |entity|
+            next if entity.corporation? && !entity.ipoed
+
+            income = entity.companies.sum(&:revenue)
+            income += calculate_synergies(entity) if entity.corporation?
```
Comment (tobymao) [inline lib/engine/game/g_rolling_stock/game.rb:196]: just an fyi when i implemented RS, doing synergies lazily is super computationally expensive

Comment (roseundy) [reply lib/engine/game/g_rolling_stock/game.rb:196]: So you recommend caching them?

File: lib/engine/game/g_rolling_stock/game.rb (hunk 1 above)
Comment (roseundy) [reply lib/engine/game/g_rolling_stock/game.rb:263]: Good point. Thanks.

File: lib/engine/game/g_rolling_stock/game.rb (hunk 2 above)
Comment (tobymao) [reply lib/engine/game/g_rolling_stock/game.rb:196]: i needed to do that when i implemented it or else it got unbearably slow

**Pull request merged**